pyriemann/sampling.py

```diff
@@ -0,0 +1,325 @@
+import numpy as np
+from functools import partial
+from pyriemann.utils.base import sqrtm
+from sklearn.utils import check_random_state
+
+
+def _pdf_r(r, sigma):
+    """pdf for the log of eigenvalues of a SPD matrix
+
+    Probability density function for the logarithm of the eigenvalues of a SPD
+    matrix samples from the Riemannian Gaussian distribution [1]_.
+
+    Parameters
+    ----------
+    r : ndarray shape (n_dim,)
+        vector defines in R^n_dim
+    sigma : float
+        dispersion of the Riemannian Gaussian distribution
+
+    Returns
+    -------
+    p : float
+        pdf applied to data point r
+
+    Notes
+    -----
+    .. versionadded:: 0.2.8.dev
+
+    References
+    ----------
+    .. [1] S. Said, L. Bombrun, Y. Berthoumieu, and J. Manton, “Riemannian
+        Gaussian distributions on the space of symmetric positive definite
+        matrices”, IEEE Trans Inf Theory, vol. 63, pp. 2153–2170, 2017.
+        https://arxiv.org/pdf/1507.01760.pdf
+    """
+
+    n_dim = len(r)
+    partial_1 = -np.sum(r**2)/sigma**2
+    partial_2 = 0
+    for i in range(n_dim):
+        for j in range(i+1, n_dim):
+            partial_2 = partial_2 + np.log(np.sinh(np.abs(r[i]-r[j])/2))
+
+    return np.exp(partial_1 + partial_2)
+
+
+def _slice_sampling(ptarget, n_samples, x0, n_burnin=20, thin=10,
+                    random_state=42):
+    """Slice sampling procedure
+
+    Implementation of a slice sampling algorithm for sampling from any target
+    pdf or a multiple of it. The implementation follows the description given
+    in page 375 of David McKay's book "Information Theory, Inference, and
+    Learning Algorithms" (2003).
```

Comment:
This could be transformed into a Reference.

Reply:
But this is also a private function, so shouldn't I leave the reference like this?

Reply:
What do you think @agramfort ?